### PR TITLE
Fix compile warnings in client app

### LINKED ISSUE #410: Warnings when compiling the client app
Running with GCC, get the following warning:
```
 [ 78%] Building C object _deps/picotls-build/CMakeFiles/cli.dir/t/cli.c.o
/home/runner/work/quicrq/quicrq/build/_deps/picotls-src/t/cli.c: In function ‘handle_connection’:
/home/runner/work/quicrq/quicrq/build/_deps/picotls-src/t/cli.c:161:35: warning: ignoring return value of ‘write’, declared with attribute warn_unused_result [-Wunused-result]
  161 |                             (void)write(sockfd, encbuf.base, encbuf.off);
      |                                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/runner/work/quicrq/quicrq/build/_deps/picotls-src/t/cli.c:170:33: warning: ignoring return value of ‘write’, declared with attribute warn_unused_result [-Wunused-result]
  170 |                                 write(1, rbuf.base, rbuf.off);
      |                                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/runner/work/quicrq/quicrq/build/_deps/picotls-src/t/cli.c:257:27: warning: ignoring return value of ‘write’, declared with attribute warn_unused_result [-Wunused-result]
  257 |                     (void)write(sockfd, wbuf.base, wbuf.off);
      |                           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

### PR DESCRIPTION
**Pull request by huitema**
Close #410

### CODE REVIEW

**Comment by huitema** [comment]
This is a very simple PR to fix some compile warnings on Linux when the user sets the "strict" modes for warnings.

**Review by kazuho — COMMENTED** [review]
Can we just add `-Wno-unused-result` instead?

GCC and clang seem to have contradicting views (see https://gcc.gnu.org/bugzilla/show_bug.cgi?id=66425 and https://bugs.llvm.org/show_bug.cgi?id=51228), but I think clang is correct here.

Cast to `void` is an expression of the intent to ignore result. The code does not need to be more complicated than that.

I'd be positive to adopting these types of changes if they were for public header files (because header files would be complied by the options that an application developer chooses), but this is about a source file that is compiled with the options written in our CMakeFiles.txt.

**Comment by huitema** [comment]
I agree that clang is doing the right thing, and that an explicit cast to (void) *should* be enough. Still, some of the companies using picotls (and picoquic) systematically use `-Wpedantic -Wextra -Wall`. These three warnings are the only one that I see, and I would be happy to see them fixed. Yes, it makes the code a bit longer, but the issue only happens when compiling `t/cli.c`. It is a test program, and I am not sure that a few extra line of code matter.

I don't like suppressing with a global option like `-Wno-unused-result`. Testing return arguments is a good practice. The examples include writing on a socket. It makes sense to actually check whether the socket broke. I could see using `pragma` to suppress checks for a specific line after verifying that there is no danger, something envisaged in the thread that you pointed at. But it is hard to write such `pragma` in a portable way -- each compiler has its own dialects. I suppose that we could define that as an "PTLS_UNCHECKED" macro, and have the macro defined according to the compiler in use in a header file. Is that what you are proposing?

**Comment by kazuho** [comment]
> Still, some of the companies using picotls (and picoquic) systematically use `-Wpedantic -Wextra -Wall`.
Am I correct to understand that they are building the cli program with their own compiler flags? cli.c is used only by the cli program, I'm not sure if there's a benefit in making such change. Doing that does not help identify issues within the code that they are writing.

That said, ...

> The examples include writing on a socket. It makes sense to actually check whether the socket broke.

I think this is actually a good point. Let me review the PR.

File: t/cli.c [hunk 1]
```diff
@@ -157,17 +157,23 @@ static int handle_connection(int sockfd, ptls_context_t *ctx, const char *server
                     } else if (ret == PTLS_ERROR_IN_PROGRESS) {
                         /* ok */
                     } else {
-                        if (encbuf.off != 0)
-                            (void)write(sockfd, encbuf.base, encbuf.off);
+                        if (encbuf.off != 0) {
+                            if (write(sockfd, encbuf.base, encbuf.off) < 0) {
+                                fprintf(stderr, "write encbuf on socket %d fails\n", (int)sockfd);
```
Comment (kazuho) [inline t/cli.c:162]: I do not think we need to say anything here.

We are merely trying to be nice by sending an Alert. But it is not uncommon for peers to just close the socket without waiting for an Alert. If that happens, that's totally fine.

But unfortunately, due to how GCC works, there's no good way of ignoring the result :-/

Therefore, I might suggest creating a wrapper function around `write`, that repeatedly calls `write` while EINTR is reported. We can call such function here and just ignore the result.

File: t/cli.c [hunk 2]
```diff
@@ -157,17 +157,23 @@ static int handle_connection(int sockfd, ptls_context_t *ctx, const char *server
...
                     if ((ret = ptls_receive(tls, &rbuf, bytebuf + off, &leftlen)) == 0) {
                         if (rbuf.off != 0) {
                             data_received += rbuf.off;
-                            if (input_file != input_file_is_benchmark)
-                                write(1, rbuf.base, rbuf.off);
+                            if (input_file != input_file_is_benchmark) {
+                                if (write(1, rbuf.base, rbuf.off) < 0) {
+                                    fprintf(stderr, "write on file 1 fails.");
+                                }
+                            }
```
Comment (kazuho) [inline t/cli.c:176]: Here, I think we have to either of the following:
* Use the aforementioned wrapper function and ignore the result, based on the perception that we can ignore write errors to STDOUT (EPIPE would kill the process).
* Use the aforementioned wrapper function and if there is an error, report that, then jump to `Exit`. We need to exit when there is an unrecoverable error.

File: t/cli.c [hunk 3]
```diff
@@ -253,8 +259,11 @@ static int handle_connection(int sockfd, ptls_context_t *ctx, const char *server
                 if ((ret = ptls_send_alert(tls, &wbuf, PTLS_ALERT_LEVEL_WARNING, PTLS_ALERT_CLOSE_NOTIFY)) != 0) {
                     fprintf(stderr, "ptls_send_alert:%d\n", ret);
                 }
-                if (wbuf.off != 0)
-                    (void)write(sockfd, wbuf.base, wbuf.off);
+                if (wbuf.off != 0) {
+                    if (write(sockfd, wbuf.base, wbuf.off) < 0) {
+                        fprintf(stderr, "write on close on socket %d fails\n", (int)sockfd);
+                    }
+                }
```
Comment (kazuho) [inline t/cli.c:266]: Here, I think we should use the aforementioned wrapper function and jump to Exit if the write failed.

File: t/cli.c (hunk 1 above)
Comment (huitema) [reply t/cli.c:162]: OK, so something like `PTLS_REPEAT_ON_EINTR(write(sockfd, encbuf.base, encbuf.off))`?

File: t/cli.c (hunk 3 above)
Comment (huitema) [reply t/cli.c:266]: That means the macro `PTLS_REPEAT_ON_EINTR` has to return -1 if `errno != EINTR`. Something like:
```
PTLS_REPEAT_ON_EINTR(ret, block) while((ret=code) == -1){if(errno != EINTR) break;}
```
Maybe defined as no-op in Windows.

Comment (huitema) [reply t/cli.c:266]: Cannot actually jump to exit directly, need to call `ptls_buffer_dispose(&wbuf);`

**Comment by huitema** [comment]
@kazuho I think this is ready. I tried to add a github action to test with `gcc` and the flags `-Wpedantic -Wextra -Wall` but I could not find an easy way to run it, as apparently `g++` is not installed in the docker build. I did test it locally, and verified that issue #410 is fixed.

**Comment by huitema** [comment]
@kazuho do you want to check this in, or shall I drop it?

**Pull request closed without merging**